### ISSUE TITLE
If a claim stays open for too long, remind the user to do it.

### ISSUE THREAD

**Issue by nic-hartley** [GrafeasGroup/tor#93]
I dunno about anyone else, but I sometimes have trouble remembering to mark a post as `done`. It would be super useful if the bot, say, replied to the `claim` comment after six hours if you hadn't marked that post as `done`.

[I've made a Gist to describe two possible ways of doing this.](https://gist.github.com/anonymous/b837c54bf791210a688c313179f8a7c0) The better solution -- the first -- will probably have to wait until the worker/queue rewrite.

**Comment by TheLonelyGhost** [comment]
Interesting... I'd thought of similar functionality as well, but wasn't sure of a good way to implement it with our current resources.

The threading model you mention in the gist I'm going to veto right now. From experience, it'll only add more problems than it'll help when using python.

I do like the queue model, but there are some complications related to concurrency here too which I'd rather rely on other software to handle. Lemme just spew some thoughts here for a second.

---

It would be trivial if we used RabbitMQ's dead-letter functionality, for instance.

It would also be easy if we polled via SQL query in a relational database, though we'd have to figure out how often we'd poll the database so it wouldn't fall over.

We're using Redis to persist private data, and it doesn't have a whole lot to make building a worker queue easy with its primitives. The closest thing is the queue vs stack behavior on top of redis' set primitive. This means sticking with Redis will require us to write our own worker queue library or to find an existing one in python. I'm much more in favor of adopting someone else's so I won't have to maintain it. 😄

**Comment by nic-hartley** [comment]
I should clarify that I didn't have any third-party stuff like Redis in mind when I wrote the gist, so there are _definitely_ other solutions out there that I just didn't consider. I don't have nearly enough experience with Redis to say anything one way or another, but that sounds good!

**Comment by xanderyzwich** [comment]
I would agree with the relational database idea.  Interaction would only be needed on post (insert), claim(update), done(delete), a fixed interval checker(query), and a delete upon post purge.

**Comment by nic-hartley** [comment]
For what it's worth, if we used SQLite rather than a separate sequel server, we could poll basically as much as we wanted -- it'd be little different to reading and writing a JSON file, but with the power of SQL and the speed of a relational database, so the speed would be limited based on what the hardware could support, rather than the server software. It would also make backups and setups much easier; we'd just have to copy and paste the database file around, which could very easily be automated with built-in tools.

**Comment by TheLonelyGhost** [comment]
Looks like this would be best suited as a separate section to the `tor` bot's routine and pull from the API that's being built. As of writing, the bot's routine is as follows:

1. Poll inbox for unread messages, processing each one and marking it as read
2. Poll each monitored subreddit for posts that make it through the configured filters

This seems like a good item to have as a third entry to its routine. Either way, we'll have to wait until the API is built for this to be reasonable.

**Comment by xanderyzwich** [comment]
I think that rolling this functionality into the existing routine to find what claims have gone unfinished for more than 6 hours at the time of running. I believe that doing this in a batch style will be fine as far as transcriber experience.  The only issue that I might want to research is any rate limiting against posting comments from the bot, although to be honest I don't expect that there is a high volume of claims that are not marked done within 6 hours.

**Comment by TheLonelyGhost** [comment]
Now that blossom is deployed, this seems like a solid enhancement. @Pf441, do you have any thoughts on this?

**Comment by TheodoreHua** [comment]
Correct me if I'm wrong but this feature is now implemented in the form of u/tor reminding the user to complete their existing transcriptions when claiming?

**Comment by TimJentzsch** [comment]
Partially, but not quite. The reminder is only sent when the limit is already reached, so if you can do 2 transcriptions at a time, you might always have 1 in progress that you don't know about.

Additionally, if a user doesn't transcribe anymore at all, they'll never get a reminder.
